zbus/src/connection/socket/unixexec.rs

```diff
@@ -0,0 +1,118 @@
+use std::io;
+use std::os::fd::BorrowedFd;
+
+#[cfg(feature = "tokio")]
+use tokio::io::{AsyncReadExt, ReadBuf};
+
+use crate::process::{Child, ChildStdin, ChildStdout};
+
+use super::{ReadHalf, RecvmsgResult, Socket, Split, WriteHalf};
+
+#[derive(Debug)]
+pub struct UnixExecStream {
+    stdin: ChildStdin,
+    stdout: ChildStdout,
+}
+
+impl UnixExecStream {
+    fn into_split(self) -> (ChildStdout, ChildStdin) {
+        (self.stdout, self.stdin)
+    }
+}
+
+impl Socket for UnixExecStream {
+    type ReadHalf = ChildStdout;
+    type WriteHalf = ChildStdin;
+
+    fn split(self) -> Split<Self::ReadHalf, Self::WriteHalf> {
+        let (read, write) = self.into_split();
+
+        Split { read, write }
+    }
+}
+
+impl TryFrom<Child> for UnixExecStream {
+    type Error = crate::Error;
+
+    fn try_from(child: Child) -> Result<Self, Self::Error> {
+        let (stdin, stdout) = child.take_stdio()?;
+
+        Ok(UnixExecStream { stdin, stdout })
+    }
+}
+
+#[async_trait::async_trait]
+impl ReadHalf for ChildStdout {
+    #[cfg(not(feature = "tokio"))]
+    async fn recvmsg(&mut self, buf: &mut [u8]) -> RecvmsgResult {
+        match futures_util::AsyncReadExt::read(&mut self.0, buf).await {
+            Err(e) => Err(e),
+            Ok(len) => {
+                #[cfg(unix)]
+                let ret = (len, vec![]);
+                #[cfg(not(unix))]
+                let ret = len;
+                Ok(ret)
+            }
+        }
+    }
+
+    #[cfg(feature = "tokio")]
+    async fn recvmsg(&mut self, buf: &mut [u8]) -> RecvmsgResult {
+        let mut read_buf = ReadBuf::new(buf);
+        self.0.read_buf(&mut read_buf).await.map(|_| {
+            let ret = read_buf.filled().len();
+            #[cfg(unix)]
+            let ret = (ret, vec![]);
+
+            ret
+        })
+    }
+}
+
+#[async_trait::async_trait]
+impl WriteHalf for ChildStdin {
+    #[cfg(not(feature = "tokio"))]
```

Comment:
If each method will have a different impl for different backend, it'd be best to feature-gate the impl rather. Please see other socket impls for an example (e.g `tcp` mod).